src/txn/txn_timestamp.c

```diff
@@ -306,3 +297,109 @@ __wt_txn_set_timestamp(WT_SESSION_IMPL *session, const char *cfg[])
 
 	return (0);
 }
+
+/*
+ * __wt_txn_set_commit_timestamp --
+ *	Publish a transaction's commit timestamp.
+ */
+void
+__wt_txn_set_commit_timestamp(WT_SESSION_IMPL *session)
+{
+	WT_TXN *prev, *txn;
+	WT_TXN_GLOBAL *txn_global;
+
+	txn = &session->txn;
+	txn_global = &S2C(session)->txn_global;
+
+	if (F_ISSET(txn, WT_TXN_HAS_TS_COMMIT))
+		return;
+
+	__wt_writelock(session, &txn_global->commit_timestamp_rwlock);
+	for (prev = TAILQ_LAST(&txn_global->commit_timestamph, __wt_txn_cts_qh);
+	    prev != NULL && __wt_timestamp_cmp(
+	    prev->commit_timestamp, txn->commit_timestamp) > 0;
+	    prev = TAILQ_PREV(prev, __wt_txn_cts_qh, commit_timestampq))
+		;
+	if (prev == NULL)
+		TAILQ_INSERT_HEAD(
+		    &txn_global->commit_timestamph, txn, commit_timestampq);
+	else
+		TAILQ_INSERT_AFTER(&txn_global->commit_timestamph,
+		    prev, txn, commit_timestampq);
+	__wt_writeunlock(session, &txn_global->commit_timestamp_rwlock);
```

Comment:
I think some statistics in here would be interesting - knowing how the queues are being used could give us a way to validate intuition. I'm happy for that to be another ticket.